### PR TITLE
Fix monadic bindings (new 4.08 syntax)

### LINKED ISSUE #907: ocamlformat generates invalid OCaml syntax
Observed with: 6d9f553ee66b7a56ed20701ae9625251b2f31957

The following OCaml file implements a state monad. I'm not sure what particular part of this snippet is causing the bug, and I'm hesitant to try minimizing it because it's possible that multiple parts of it are triggering the issue.

```ocaml
module Make (K : sig type t end) () = struct
  type state = K.t

  type 'a t = state -> 'a * state

  let run s t =
    t s

  let pure a = fun s -> a, s

  let get = fun s -> s, s

  let put s = fun _ -> (), s

  let modify f =
    fun s -> (), f s

  module Syntax = struct
    let ( let* ) t f =
      fun s ->
        let a, s2 = t s in
        (f a) s2

    let ( and* ) t1 t2 =
       fun s ->
         let a, s2 = t1 s in
         let b, s3 = t2 s2 in
         (a, b), s3
  end

  let map f t =
    let open Syntax in
    let* a = t in
    pure (f a)
end
```

`.ocamlformat`:

```
profile = conventional
```

(I've also tried the `sparse` and `ocamlformat` profiles, and the result is the same.)

```bash
$ ocamlformat state.ml
ocamlformat: Cannot process "state.ml".
  Please report this bug at https://github.com/ocaml-ppx/ocamlformat/issues.
  BUG: generating invalid ocaml syntax.
```

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #907
The rules for the indent names are from https://github.com/ocaml/ocaml/pull/1947

@hakuch feel free to give us some feedback before we merge this branch.

### CODE REVIEW

File: src/Ast.ml [hunk 1]
```diff
@@ -53,7 +58,13 @@ let is_infix_id i =
     , ( "!=" | "land" | "lor" | "lxor" | "mod" | "::" | ":=" | "asr" | "lsl"
       | "lsr" | "or" | "||" ) ) ->
       true
-  | _ -> false
+  | _ ->
+      String.length i > 3
+      && ( String.is_substring i ~substring:"let"
```
Comment (hhugo) [inline src/Ast.ml]: Should this be String.is_prefix?

Comment (gpetiot) [reply src/Ast.ml]: Good catch, fixed.

**Comment by hakuch** [comment]
I checked out this branch locally, and the snippet in #907 still fails, but this time with a different error.

I ran `ocamlformat --debug state.ml` and this is the output:

```
FAIL
Exp:
( and* ) t1 t2

expression (state.ml[37,704+6]..[37,704+20])
  Pexp_apply
  expression (
FAIL
Exp:
( and* ) t1 t2

expression (state.ml[37,704+6]..[37,704+20])
  Pexp_apply
  expression (state.ml[37,704+6]..[37,704+14])
    Pexp_ident "and*" (state.ml[37,704+6]..[37,704+14])
  [
    <arg>
    Nolabel
      expression (state.ml[37,704+15]..[37,704+17])
        Pexp_ident "t1" (state.ml[37,704+15]..[37,704+17])
    <arg>
    Nolabel
      expression (state.ml[37,704+18]..[37,704+20])
        Pexp_ident "t2" (state.ml[37,704+18]..[37,704+20])
  ]

Raised by primitive operation at file "src/bytes0.ml" (inlined), line 45, characters 22-39
```

Lots more output follows, and then

```
ocamlformat: Cannot process "state.ml".
  Please report this bug at https://github.com/ocaml-ppx/ocamlformat/issues.
  BUG: unhandled exception. Use [--debug] for details.
("Stack overflow")
```

**Comment by gpetiot** [comment]
Everything should be fixed now.

**Comment by hakuch** [comment]
I should be able to check again in 8 to 12 hours from now. Thanks!

File: src/Ast.ml [hunk 2]
```diff
@@ -44,6 +44,23 @@ let is_prefix exp =
   | Pexp_ident {txt= Lident i; _} -> is_prefix_id i
   | _ -> false
 
+let is_infix_symbol = function
+  | '$' | '&' | '*' | '+' | '-' | '/' | '<' | '=' | '>' | '@' | '^' | '|' ->
```
Comment (hhugo) [inline src/Ast.ml]: I think `is_infix_symbol` should be renamed to something more specific (is_kwdopchar ?). Also, I don't think this list is accurate. `!`, `%`, `:`, `?` seem to be missing.

File: src/Ast.ml [hunk 3]
```diff
@@ -44,6 +44,23 @@ let is_prefix exp =
...
+      true
+  | _ -> false
+
+let is_monadic_binding_id s =
+  String.length s > 3
+  && (String.is_prefix s ~prefix:"let" || String.is_prefix s ~prefix:"and")
+  && String.for_all
+       (String.sub s ~pos:3 ~len:(String.length s - 3))
```
Comment (hhugo) [inline src/Ast.ml]: Can we find a way to not copy the string ?

File: src/Ast.ml (hunk 2 above)
Comment (hhugo) [reply src/Ast.ml]: It should consider the union of the following two definition:
```
let dotsymbolchar =
  ['!' '$' '%' '&' '*' '+' '-' '/' ':' '=' '>' '?' '@' '^' '|']
let kwdopchar =
  ['$' '&' '*' '+' '-' '/' '<' '=' '>' '@' '^' '|']
```

Comment (gpetiot) [reply src/Ast.ml]: > I think `is_infix_symbol` should be renamed to something more specific (is_kwdopchar ?). Also, I don't think this list is accurate. `!`, `%`, `:`, `?` seem to be missing.

This list comes from the ocaml PR: https://github.com/ocaml/ocaml/pull/1947/files#diff-635f3857747d984a68e56f2afdc27a4b

**Comment by hakuch** [comment]
Apologies for the delay. With the latest version of this PR, all my usages of the monadic and applicative `let` bindings are handled correctly by ocamlformat!

File: src/Ast.ml (hunk 3 above)
Comment (gpetiot) [reply src/Ast.ml]: Using `lfindi` now, it should be better.

File: src/Ast.ml (hunk 2 above)
Comment (gpetiot) [reply src/Ast.ml]: Good catch, fixed.

**Pull request merged**